<a href="https://colab.research.google.com/github/martonhajdu/martonhajdu-DataScience-GenAI-Submissions/blob/main/Seminar_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
from google.colab import files
upload = files.upload()

Saving students_two.csv to students_two.csv


In [ ]:
import pandas as pd
df = pd.read_csv("students_two.csv")
df.head()

,Id,Student_Age,High_School_Type,Scholarship,Additional_Work,Sports_activity,Transportation,Weekly_Study_Hours,Attendance,Reading,Notes,Listening_in_Class,Project_work,Grade
0,5001,21,Other,50%,Yes,No,Private,0,Always,Yes,Yes,No,No,AA
1,5002,20,Other,50%,Yes,No,Private,0,Always,Yes,No,Yes,Yes,AA
2,5003,21,State,50%,No,No,Private,2,Never,No,No,No,Yes,AA
3,5004,18,Private,50%,Yes,No,Bus,2,Always,No,Yes,No,No,AA
4,5005,22,Private,50%,No,No,Bus,12,Always,Yes,No,Yes,Yes,AA


In [ ]:
df.isnull().sum()

,0
Id,0
Student_Age,0
High_School_Type,0
Scholarship,1
Additional_Work,0
Sports_activity,0
Transportation,0
Weekly_Study_Hours,0
Attendance,0
Reading,0


In [ ]:
df.dropna(inplace=True)
df.isnull().sum()

,0
Id,0
Student_Age,0
High_School_Type,0
Scholarship,0
Additional_Work,0
Sports_activity,0
Transportation,0
Weekly_Study_Hours,0
Attendance,0
Reading,0


In [ ]:
df = df.drop(['Id'], axis = 1)
df.head()

,Student_Age,High_School_Type,Scholarship,Additional_Work,Sports_activity,Transportation,Weekly_Study_Hours,Attendance,Reading,Notes,Listening_in_Class,Project_work,Grade
0,21,Other,50%,Yes,No,Private,0,Always,Yes,Yes,No,No,AA
1,20,Other,50%,Yes,No,Private,0,Always,Yes,No,Yes,Yes,AA
2,21,State,50%,No,No,Private,2,Never,No,No,No,Yes,AA
3,18,Private,50%,Yes,No,Bus,2,Always,No,Yes,No,No,AA
4,22,Private,50%,No,No,Bus,12,Always,Yes,No,Yes,Yes,AA


In [ ]:
# Converting features into numbers, so we can use all data for modeling
df = pd.get_dummies(df, columns=['High_School_Type'], drop_first=True, dtype=int)
df.head()

# Converted high school type into Private and State, we don't need a third column for other, if both are 0,
# we know it's in the other school type


,Student_Age,Scholarship,Additional_Work,Sports_activity,Transportation,Weekly_Study_Hours,Attendance,Reading,Notes,Listening_in_Class,Project_work,Grade,High_School_Type_Private,High_School_Type_State
0,21,50%,Yes,No,Private,0,Always,Yes,Yes,No,No,AA,0,0
1,20,50%,Yes,No,Private,0,Always,Yes,No,Yes,Yes,AA,0,0
2,21,50%,No,No,Private,2,Never,No,No,No,Yes,AA,0,1
3,18,50%,Yes,No,Bus,2,Always,No,Yes,No,No,AA,1,0
4,22,50%,No,No,Bus,12,Always,Yes,No,Yes,Yes,AA,1,0


In [ ]:
# Converting Scholarship to a number
df['Scholarship'] = df['Scholarship'].str.replace('%', '').astype(float) / 100
df.head()

,Student_Age,Scholarship,Additional_Work,Sports_activity,Transportation,Weekly_Study_Hours,Attendance,Reading,Notes,Listening_in_Class,Project_work,Grade,High_School_Type_Private,High_School_Type_State
0,21,0.5,Yes,No,Private,0,Always,Yes,Yes,No,No,AA,0,0
1,20,0.5,Yes,No,Private,0,Always,Yes,No,Yes,Yes,AA,0,0
2,21,0.5,No,No,Private,2,Never,No,No,No,Yes,AA,0,1
3,18,0.5,Yes,No,Bus,2,Always,No,Yes,No,No,AA,1,0
4,22,0.5,No,No,Bus,12,Always,Yes,No,Yes,Yes,AA,1,0


In [ ]:
# Converting 'Additional work'
df['Additional_Work'] = df['Additional_Work'].map({'Yes': 1, 'No': 0})
df.head()

,Student_Age,Scholarship,Additional_Work,Sports_activity,Transportation,Weekly_Study_Hours,Attendance,Reading,Notes,Listening_in_Class,Project_work,Grade,High_School_Type_Private,High_School_Type_State
0,21,0.5,1,No,Private,0,Always,Yes,Yes,No,No,AA,0,0
1,20,0.5,1,No,Private,0,Always,Yes,No,Yes,Yes,AA,0,0
2,21,0.5,0,No,Private,2,Never,No,No,No,Yes,AA,0,1
3,18,0.5,1,No,Bus,2,Always,No,Yes,No,No,AA,1,0
4,22,0.5,0,No,Bus,12,Always,Yes,No,Yes,Yes,AA,1,0


In [ ]:
# Checking for the values of these columns
categorical_cols = ['Sports_activity', 'Transportation', 'Attendance', 'Reading', 'Notes', 'Listening_in_Class', 'Project_work', 'Grade']

for col in categorical_cols:
    print(f"Unique values for '{col}': {df[col].unique()}")

Unique values for 'Sports_activity': ['No' 'Yes']
Unique values for 'Transportation': ['Private' 'Bus']
Unique values for 'Attendance': ['Always' 'Never' 'Sometimes' '3']
Unique values for 'Reading': ['Yes' 'No']
Unique values for 'Notes': ['Yes' 'No' '6']
Unique values for 'Listening_in_Class': ['No' 'Yes' '6']
Unique values for 'Project_work': ['No' 'Yes']
Unique values for 'Grade': ['AA' 'BA' 'CC' 'Fail' 'BB' 'CB' 'DD' 'DC']


In [ ]:
# Dropping the rows where 6 appears in 'Notes' and 'Listening_in_Class' and dropping the 3 from 'Attendance'
df = df[(df['Notes'] != '6') & (df['Listening_in_Class'] != '6') & (df['Attendance'] != '3')]
df.head()

,Student_Age,Scholarship,Additional_Work,Sports_activity,Transportation,Weekly_Study_Hours,Attendance,Reading,Notes,Listening_in_Class,Project_work,Grade,High_School_Type_Private,High_School_Type_State
0,21,0.5,1,No,Private,0,Always,Yes,Yes,No,No,AA,0,0
1,20,0.5,1,No,Private,0,Always,Yes,No,Yes,Yes,AA,0,0
2,21,0.5,0,No,Private,2,Never,No,No,No,Yes,AA,0,1
3,18,0.5,1,No,Bus,2,Always,No,Yes,No,No,AA,1,0
4,22,0.5,0,No,Bus,12,Always,Yes,No,Yes,Yes,AA,1,0


In [ ]:
# Convertting 'Yes' and 'No' to 1 and 0
binary_cols = ['Sports_activity', 'Reading', 'Notes', 'Listening_in_Class', 'Project_work']

for col in binary_cols:
    df[col] = df[col].map({'Yes': 1, 'No': 0})

df.head()

,Student_Age,Scholarship,Additional_Work,Sports_activity,Transportation,Weekly_Study_Hours,Attendance,Reading,Notes,Listening_in_Class,Project_work,Grade,High_School_Type_Private,High_School_Type_State
0,21,0.5,1,0,Private,0,Always,1,1,0,0,AA,0,0
1,20,0.5,1,0,Private,0,Always,1,0,1,1,AA,0,0
2,21,0.5,0,0,Private,2,Never,0,0,0,1,AA,0,1
3,18,0.5,1,0,Bus,2,Always,0,1,0,0,AA,1,0
4,22,0.5,0,0,Bus,12,Always,1,0,1,1,AA,1,0


In [ ]:
# Same for 'Transportation'
df['Transportation'] = df['Transportation'].map({'Private': 1, 'Bus': 0})
df.head()

,Student_Age,Scholarship,Additional_Work,Sports_activity,Transportation,Weekly_Study_Hours,Attendance,Reading,Notes,Listening_in_Class,Project_work,Grade,High_School_Type_Private,High_School_Type_State
0,21,0.5,1,0,1,0,Always,1,1,0,0,AA,0,0
1,20,0.5,1,0,1,0,Always,1,0,1,1,AA,0,0
2,21,0.5,0,0,1,2,Never,0,0,0,1,AA,0,1
3,18,0.5,1,0,0,2,Always,0,1,0,0,AA,1,0
4,22,0.5,0,0,0,12,Always,1,0,1,1,AA,1,0


In [ ]:
# For 'Attendance' a range makes sense
df['Attendance'] = df['Attendance'].map({'Never': 0, 'Sometimes': 1, 'Always': 2})
df.head()

,Student_Age,Scholarship,Additional_Work,Sports_activity,Transportation,Weekly_Study_Hours,Attendance,Reading,Notes,Listening_in_Class,Project_work,Grade,High_School_Type_Private,High_School_Type_State
0,21,0.5,1,0,1,0,2,1,1,0,0,AA,0,0
1,20,0.5,1,0,1,0,2,1,0,1,1,AA,0,0
2,21,0.5,0,0,1,2,0,0,0,0,1,AA,0,1
3,18,0.5,1,0,0,2,2,0,1,0,0,AA,1,0
4,22,0.5,0,0,0,12,2,1,0,1,1,AA,1,0


In [ ]:
# Converting the 'Grade' column. Fail will be 0, rest (Pass) will be 1
df['Grade'] = df['Grade'].apply(lambda x: 0 if x == 'Fail' or 'D' in x else 1)
df.head()

,Student_Age,Scholarship,Additional_Work,Sports_activity,Transportation,Weekly_Study_Hours,Attendance,Reading,Notes,Listening_in_Class,Project_work,Grade,High_School_Type_Private,High_School_Type_State
0,21,0.5,1,0,1,0,2,1,1,0,0,1,0,0
1,20,0.5,1,0,1,0,2,1,0,1,1,1,0,0
2,21,0.5,0,0,1,2,0,0,0,0,1,1,0,1
3,18,0.5,1,0,0,2,2,0,1,0,0,1,1,0
4,22,0.5,0,0,0,12,2,1,0,1,1,1,1,0


In [ ]:
Y_values = df[['Grade']].values.ravel() # change to list
Y_values

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0,
       1, 0, 0, 1, 1, 1, 1, 1])

In [ ]:
# Dropping 'Grade'. We will try to predict it, so it doesn't need to be part of the training data
X_values = df.drop(['Grade'], axis=1)
X_values.head()

,Student_Age,Scholarship,Additional_Work,Sports_activity,Transportation,Weekly_Study_Hours,Attendance,Reading,Notes,Listening_in_Class,Project_work,High_School_Type_Private,High_School_Type_State
0,21,0.5,1,0,1,0,2,1,1,0,0,0,0
1,20,0.5,1,0,1,0,2,1,0,1,1,0,0
2,21,0.5,0,0,1,2,0,0,0,0,1,0,1
3,18,0.5,1,0,0,2,2,0,1,0,0,1,0
4,22,0.5,0,0,0,12,2,1,0,1,1,1,0


In [ ]:
# Scalign X
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
# fit and transform the data and convert back to DataFrame
X_values = pd.DataFrame(scaler.fit_transform(X_values), columns=X_values.columns)
X_values

,Student_Age,Scholarship,Additional_Work,Sports_activity,Transportation,Weekly_Study_Hours,Attendance,Reading,Notes,Listening_in_Class,Project_work,High_School_Type_Private,High_School_Type_State
0,0.375,0.333333,1.0,0.0,1.0,0.000000,1.0,1.0,1.0,0.0,0.0,0.0,0.0
1,0.250,0.333333,1.0,0.0,1.0,0.000000,1.0,1.0,0.0,1.0,1.0,0.0,0.0
2,0.375,0.333333,0.0,0.0,1.0,0.166667,0.0,0.0,0.0,0.0,1.0,0.0,1.0
3,0.000,0.333333,1.0,0.0,0.0,0.166667,1.0,0.0,1.0,0.0,0.0,1.0,0.0
4,0.500,0.333333,0.0,0.0,0.0,1.000000,1.0,1.0,0.0,1.0,1.0,1.0,0.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
135,0.500,0.333333,1.0,1.0,1.0,0.000000,1.0,0.0,1.0,0.0,1.0,0.0,1.0
136,0.000,0.666667,0.0,0.0,1.0,0.000000,0.0,0.0,1.0,1.0,0.0,0.0,1.0
137,0.000,0.666667,0.0,0.0,1.0,0.000000,1.0,1.0,0.0,0.0,0.0,1.0,0.0
138,0.500,0.666667,1.0,1.0,0.0,1.000000,0.5,0.0,1.0,0.0,1.0,0.0,1.0


In [ ]:
# Training Test split
from sklearn.model_selection  import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X_values, Y_values, test_size = 0.2, random_state=1984, stratify=Y_values)

# print the shapes to check everything is OK
print(X_train.shape)
print(X_test.shape)
print(Y_train.shape)
print(Y_test.shape)

(112, 13)
(28, 13)
(112,)
(28,)


In [ ]:
# This is a classification problem
# We will use Logistic Regression for that
from sklearn.linear_model import LogisticRegression

# Create the Logistic Regression model
model = LogisticRegression()

# Train the model using the training data
model.fit(X_train, Y_train)

LogisticRegression()

In [ ]:
# Make predictions on the test data
Y_pred = model.predict(X_test)

# Print the predictions
print(Y_pred)

[1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 0 1 1 1 1 0 0 1 1 1 1 1]


In [ ]:
from sklearn.metrics import accuracy_score

# Calculate the accuracy of the model
accuracy = accuracy_score(Y_test, Y_pred)

# Print the accuracy
print(f"Accuracy: {accuracy}")

Accuracy: 0.7142857142857143


In [ ]:
from sklearn.tree import DecisionTreeClassifier

# Create the Decision Tree model
model_dt = DecisionTreeClassifier()

# Train the model using the training data
model_dt.fit(X_train, Y_train)

DecisionTreeClassifier()

In [ ]:
Y_pred_dt = model_dt.predict(X_test)

from sklearn.metrics import accuracy_score

accuracy_dt = accuracy_score(Y_test, Y_pred_dt)

print(f"Decision Tree Accuracy: {accuracy_dt}")

Decision Tree Accuracy: 0.7142857142857143


In [ ]:
from sklearn.ensemble import GradientBoostingClassifier

# Instantiate a Gradient Boosting Classifier object
model_gb = GradientBoostingClassifier()

# Train the Gradient Boosting model
model_gb.fit(X_train, Y_train)

GradientBoostingClassifier()

In [ ]:
# Make predictions on the test data
Y_pred_gb = model_gb.predict(X_test)

# Calculate the accuracy of the Gradient Boosting model
accuracy_gb = accuracy_score(Y_test, Y_pred_gb)

# Print the accuracy
print(f"Gradient Boosting Accuracy: {accuracy_gb}")

Gradient Boosting Accuracy: 0.6785714285714286


In [ ]:
print(f"Logistic Regression Accuracy: {accuracy}")
print(f"Decision Tree Accuracy: {accuracy_dt}")
print(f"Gradient Boosting Accuracy: {accuracy_gb}")

# Identify the best performing model
model_accuracies = {"Logistic Regression": accuracy, "Decision Tree": accuracy_dt, "Gradient Boosting": accuracy_gb}
best_model = max(model_accuracies, key=model_accuracies.get)
print(f"\nBased on accuracy, the best performing model is: {best_model}")

Logistic Regression Accuracy: 0.7142857142857143
Decision Tree Accuracy: 0.7142857142857143
Gradient Boosting Accuracy: 0.6785714285714286

Based on accuracy, the best performing model is: Logistic Regression


## Summary:

### Data Analysis Key Findings

*   The Decision Tree model achieved an accuracy of approximately 0.7143 on the test dataset.
*   The Gradient Boosting model achieved an accuracy of approximately 0.6786 on the test dataset.
*   Based on accuracy, both Logistic Regression and Decision Tree models performed equally well with an accuracy of approximately 0.714, which is better than the Gradient Boosting model's accuracy of approximately 0.679.

### Insights or Next Steps

*   Since Logistic Regression and Decision Tree models performed similarly in terms of accuracy, further evaluation using other metrics (e.g., precision, recall, F1-score) could provide more insights into their performance characteristics, especially if the dataset is imbalanced.
*   Hyperparameter tuning for each model, especially for Gradient Boosting which can be sensitive to its parameters, might lead to improved performance and potentially change the ranking of the models.


In [ ]:
from sklearn.model_selection import ParameterGrid

param_grid_dt = {
    'max_depth': [None, 5, 10, 15],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

param_grid_gb = {
    'n_estimators': [50, 100, 200],
    'learning_rate': [0.01, 0.1, 0.2],
    'max_depth': [3, 5, 7]
}

In [ ]:
from sklearn.model_selection import GridSearchCV
from sklearn.tree import DecisionTreeClassifier

# Instantiate a DecisionTreeClassifier object
dt = DecisionTreeClassifier()

# Instantiate GridSearchCV with the Decision Tree classifier and the parameter grid
grid_search_dt = GridSearchCV(dt, param_grid_dt, cv=5)

# Fit the GridSearchCV object to the training data
grid_search_dt.fit(X_train, Y_train)

# Print the best parameters found by GridSearchCV
print("Best parameters for Decision Tree:", grid_search_dt.best_params_)

Best parameters for Decision Tree: {'max_depth': None, 'min_samples_leaf': 4, 'min_samples_split': 10}


In [ ]:
from sklearn.metrics import accuracy_score

# Get the best Decision Tree model from the grid search
best_dt_model = grid_search_dt.best_estimator_

# Make predictions on the test data using the best model
Y_pred_tuned_dt = best_dt_model.predict(X_test)

# Calculate the accuracy of the tuned Decision Tree model
accuracy_tuned_dt = accuracy_score(Y_test, Y_pred_tuned_dt)

# Print the accuracy
print(f"Tuned Decision Tree Accuracy: {accuracy_tuned_dt}")

Tuned Decision Tree Accuracy: 0.6428571428571429


In [ ]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.model_selection import GridSearchCV

# Instantiate a GradientBoostingClassifier object
gb = GradientBoostingClassifier()

# Instantiate GridSearchCV with the Gradient Boosting classifier and the parameter grid
grid_search_gb = GridSearchCV(gb, param_grid_gb, cv=5)

# Fit the GridSearchCV object to the training data
grid_search_gb.fit(X_train, Y_train)

# Print the best parameters found by GridSearchCV
print("Best parameters for Gradient Boosting:", grid_search_gb.best_params_)

Best parameters for Gradient Boosting: {'learning_rate': 0.01, 'max_depth': 5, 'n_estimators': 100}


In [ ]:
# Get the best Gradient Boosting model from the grid search
best_gb_model = grid_search_gb.best_estimator_

# Make predictions on the test data using the best model
Y_pred_tuned_gb = best_gb_model.predict(X_test)

# Calculate the accuracy of the tuned Gradient Boosting model
accuracy_tuned_gb = accuracy_score(Y_test, Y_pred_tuned_gb)

# Print the accuracy
print(f"Tuned Gradient Boosting Accuracy: {accuracy_tuned_gb}")

Tuned Gradient Boosting Accuracy: 0.6785714285714286


In [ ]:
# Create a dictionary to store the accuracy of each model
model_accuracies_tuned = {
    "Logistic Regression": accuracy,
    "Tuned Decision Tree": accuracy_tuned_dt,
    "Tuned Gradient Boosting": accuracy_tuned_gb
}

# Print the accuracy of each model from the dictionary
print("Model Accuracies (After Tuning):")
for model_name, acc in model_accuracies_tuned.items():
    print(f"{model_name}: {acc}")

# Identify and print the name of the model with the highest accuracy
best_model_tuned = max(model_accuracies_tuned, key=model_accuracies_tuned.get)
print(f"\nBased on accuracy after tuning, the best performing model is: {best_model_tuned}")

Model Accuracies (After Tuning):
Logistic Regression: 0.7142857142857143
Tuned Decision Tree: 0.6428571428571429
Tuned Gradient Boosting: 0.6785714285714286

Based on accuracy after tuning, the best performing model is: Logistic Regression


## Summary:

### Data Analysis Key Findings

*   The best hyperparameters found for the Decision Tree model using GridSearchCV were `{'max_depth': None, 'min_samples_leaf': 4, 'min_samples_split': 10}`.
*   The accuracy of the tuned Decision Tree model on the test set is approximately 0.643.
*   The best hyperparameters found for the Gradient Boosting model using GridSearchCV were `{'learning_rate': 0.01, 'max_depth': 5, 'n_estimators': 100}`.
*   The accuracy of the tuned Gradient Boosting model on the test set is approximately 0.679.
*   Comparing the initial Logistic Regression model (accuracy: 0.714), the tuned Decision Tree model (accuracy: 0.643), and the tuned Gradient Boosting model (accuracy: 0.679), the Logistic Regression model achieved the highest accuracy.

### Insights or Next Steps

*   Despite hyperparameter tuning, neither the Decision Tree nor the Gradient Boosting models surpassed the performance of the initial Logistic Regression model on this dataset.
*   Further investigation into feature engineering or exploring other models might be beneficial to potentially improve performance beyond the Logistic Regression baseline.


# Task
Implement and tune a Random Forest model, and compare its performance to the previously implemented Logistic Regression and Gradient Boosting models for predicting student grades.

## Implement random forest

### Subtask:
Train a Random Forest model on the training data.


**Reasoning**:
Train a Random Forest model to predict student grades.



In [ ]:
from sklearn.ensemble import RandomForestClassifier

# Instantiate a RandomForestClassifier object with default parameters
model_rf = RandomForestClassifier()

# Train the Random Forest model using the training data
model_rf.fit(X_train, Y_train)

RandomForestClassifier()

## Define parameter grid for random forest

### Subtask:
Define the parameter grid for hyperparameter tuning of the Random Forest model.


**Reasoning**:
Define the parameter grid for the Random Forest model and print it.



In [ ]:
param_grid_rf = {
    'n_estimators': [50, 100, 200],
    'max_depth': [None, 5, 10, 15],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}
print(param_grid_rf)

{'n_estimators': [50, 100, 200], 'max_depth': [None, 5, 10, 15], 'min_samples_split': [2, 5, 10], 'min_samples_leaf': [1, 2, 4]}


## Perform gridsearchcv for random forest

### Subtask:
Use `GridSearchCV` to find the best hyperparameters for the Random Forest model.


**Reasoning**:
The subtask requires using GridSearchCV to find the best hyperparameters for the Random Forest model. This involves instantiating the model, setting up GridSearchCV with the defined parameter grid, and fitting it to the training data. The best parameters are then printed.



In [ ]:
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier

# Instantiate a RandomForestClassifier object
rf = RandomForestClassifier()

# Instantiate GridSearchCV with the Random Forest classifier and the parameter grid
grid_search_rf = GridSearchCV(rf, param_grid_rf, cv=5)

# Fit the GridSearchCV object to the training data
grid_search_rf.fit(X_train, Y_train)

# Print the best parameters found by GridSearchCV
print("Best parameters for Random Forest:", grid_search_rf.best_params_)

Best parameters for Random Forest: {'max_depth': 10, 'min_samples_leaf': 2, 'min_samples_split': 2, 'n_estimators': 100}


## Evaluate tuned random forest

### Subtask:
Evaluate the performance of the Random Forest model with the best hyperparameters.


**Reasoning**:
Evaluate the performance of the tuned Random Forest model using the test data and calculate its accuracy.



In [ ]:
from sklearn.metrics import accuracy_score

# Get the best Random Forest model from the grid search
best_rf_model = grid_search_rf.best_estimator_

# Make predictions on the test data using the best model
Y_pred_tuned_rf = best_rf_model.predict(X_test)

# Calculate the accuracy of the tuned Random Forest model
accuracy_tuned_rf = accuracy_score(Y_test, Y_pred_tuned_rf)

# Print the accuracy
print(f"Tuned Random Forest Accuracy: {accuracy_tuned_rf}")

Tuned Random Forest Accuracy: 0.6428571428571429


## Compare models

### Subtask:
Compare the performance of Logistic Regression, the tuned Random Forest, and the tuned Gradient Boosting models.


**Reasoning**:
Print the accuracy of the three models and compare them.



In [ ]:
# Print the accuracy of each model
print(f"Logistic Regression Accuracy: {accuracy}")
print(f"Tuned Random Forest Accuracy: {accuracy_tuned_rf}")
print(f"Tuned Gradient Boosting Accuracy: {accuracy_tuned_gb}")

# Create a dictionary to store the accuracy of each model
model_accuracies_tuned = {
    "Logistic Regression": accuracy,
    "Tuned Random Forest": accuracy_tuned_rf,
    "Tuned Gradient Boosting": accuracy_tuned_gb
}

# Print the dictionary of model accuracies
print("\nModel Accuracies (After Tuning):")
for model_name, acc in model_accuracies_tuned.items():
    print(f"{model_name}: {acc}")

# Identify and print the name of the model with the highest accuracy
best_model_tuned = max(model_accuracies_tuned, key=model_accuracies_tuned.get)
print(f"\nBased on accuracy after tuning, the best performing model is: {best_model_tuned}")

Logistic Regression Accuracy: 0.7142857142857143
Tuned Random Forest Accuracy: 0.6428571428571429
Tuned Gradient Boosting Accuracy: 0.6785714285714286

Model Accuracies (After Tuning):
Logistic Regression: 0.7142857142857143
Tuned Random Forest: 0.6428571428571429
Tuned Gradient Boosting: 0.6785714285714286

Based on accuracy after tuning, the best performing model is: Logistic Regression


## Summary:

### Data Analysis Key Findings

*   The best hyperparameters found for the Random Forest model using `GridSearchCV` were `{'max_depth': 10, 'min_samples_leaf': 2, 'min_samples_split': 2, 'n_estimators': 100}`.
*   The tuned Random Forest model achieved an accuracy of approximately 0.643 on the test set.
*   Comparing the models after tuning, Logistic Regression had the highest accuracy (0.714), followed by Tuned Gradient Boosting (0.679), and then Tuned Random Forest (0.643).

### Insights or Next Steps

*   Logistic Regression appears to be the most effective model among the three for this specific dataset and task, based on accuracy.
*   Further investigation into feature engineering or exploring other model types could potentially improve predictive performance.
